controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java

```diff
@@ -0,0 +1,92 @@
+
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final ScheduledExecutorService executor;
+
+    public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
+                                 final StreamMetadataStore streamMetaStore,
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetaStore);
+        Preconditions.checkNotNull(streamMetaTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataStore = streamMetaStore;
+        this.streamMetadataTasks = streamMetaTasks;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(final CreateReaderGroupEvent request) {
+        String scope = request.getScopeName();
+        String readerGroup = request.getRgName();
+        long requestId = request.getRequestId();
+
+        final RGOperationContext context = streamMetadataStore.createRGContext(scope, readerGroup);
+        log.info("Processing create request.");
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getVersionedReaderGroupState(scope, readerGroup,
+                true, context, executor)
+                .thenCompose(state -> {
+                    if (state.getObject().equals(ReaderGroupState.CREATING)) {
+                        String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
+                        return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
+                               .thenCompose(configRecord -> {
+                               log.info("Adding metadata to Streams");
+                               if (!ReaderGroupConfig.StreamDataRetention.values()[configRecord.getObject().getRetentionTypeOrdinal()]
+                                   .equals(ReaderGroupConfig.StreamDataRetention.NONE)) {
+                                   // update Stream metadata tables, if RG is a Subscriber
+                                   Iterator<String> streamIter = configRecord.getObject().getStartingStreamCuts().keySet().iterator();
+                                   return Futures.loop(() -> streamIter.hasNext(), () -> {
+                                          Stream stream = Stream.of(streamIter.next());
+                                          return streamMetadataStore.addSubscriber(stream.getScope(),
+                                                       stream.getStreamName(), scopedRGName, null, executor);
+                                      }, executor);
+                                   }
+                               return CompletableFuture.completedFuture(null);
+                               }).thenCompose(v ->
+                                  Futures.toVoid(streamMetadataStore.createStream(scope, NameUtils.getStreamForReaderGroup(readerGroup),
```

Comment:
this will only create the metadata for the rg stream.. it is not creating the segments.. 
i suggest you invoke createRg stream method on streamMetadataTasks to do this.

also since create for the RG stream can also come from the users, so there can be locking contentions for creation of the stream. but its not a problem per se except that this event's processing can be delayed until the stream gets created outside this context.
however, older clients will continue to create RG streams directly.. newer clients create rg streams using the new createRG api. 
so there shouldnt be a real contention of the sort.

Reply:
Ok. Yes sure I'll add createRGStream() to streamMetadataTasks and invoke that.

Reply:
fixed